### PR TITLE
fix build for musl 1.2.3

### LINKED ISSUE #67763: Build of coreclr fails with musl 1.2.3
### Description

Build of coreclr fails with musl 1.2.3 due to following error:
```
      [ 14%] Building C object Native.Unix/System.Security.Cryptography.Native/CMakeFiles/objlib.dir/pal_x509_root.c.o
      /home/buildozer/aports/community/dotnet6-stage0/src/source-build-tarball-6.0.102/src/runtime.839cdfb0ecca5e0be3dbccd926e7651ef50fdf10/artifacts/source-build/self/src/src/coreclr/pal/src/init/pal.cpp:350:24: error: no matching function for call to 'InterlockedCompareExchangePointerT'
                  if(NULL != InterlockedCompareExchangePointer(&init_critsec, &temp_critsec, NULL))
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      /home/buildozer/aports/community/dotnet6-stage0/src/source-build-tarball-6.0.102/src/runtime.839cdfb0ecca5e0be3dbccd926e7651ef50fdf10/artifacts/source-build/self/src/src/coreclr/pal/src/include/pal/palinternal.h:727:43: note: expanded from macro 'InterlockedCompareExchangePointer'
      #define InterlockedCompareExchangePointer InterlockedCompareExchangePointerT
                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      /home/buildozer/aports/community/dotnet6-stage0/src/source-build-tarball-6.0.102/src/runtime.839cdfb0ecca5e0be3dbccd926e7651ef50fdf10/artifacts/source-build/self/src/src/coreclr/pal/src/include/pal/palinternal.h:715:11: note: candidate function template not viable: no known conversion from 'nullptr_t' to 'int' for 3rd argument
      inline T* InterlockedCompareExchangePointerT(
                ^
      /home/buildozer/aports/community/dotnet6-stage0/src/source-build-tarball-6.0.102/src/runtime.839cdfb0ecca5e0be3dbccd926e7651ef50fdf10/artifacts/source-build/self/src/src/coreclr/pal/src/include/pal/palinternal.h:684:4: note: candidate template ignored: could not match 'T *' against 'nullptr_t'
      T* InterlockedCompareExchangePointerT(
         ^
      /home/buildozer/aports/community/dotnet6-stage0/src/source-build-tarball-6.0.102/src/runtime.839cdfb0ecca5e0be3dbccd926e7651ef50fdf10/artifacts/source-build/self/src/src/coreclr/pal/src/include/pal/palinternal.h:705:11: note: candidate template ignored: could not match 'T *' against 'nullptr_t'
      inline T* InterlockedCompareExchangePointerT(
                ^
      1 error generated.
```

Caused by the following change in musl:
https://git.musl-libc.org/cgit/musl/commit?id=98e688a9da5e7b2925dda17a2d6820dddf1fb287

### Reproduction Steps

1) Setup Alpine Edge environment with musl 1.2.3.
2) Attempt a build

### Expected behavior

Build should pass, as with musl 1.2.2

### Actual behavior

Failure, as described in description.

### Regression?

_No response_

### Known Workarounds

Removing the problematic commit from musl does the trick.

### Configuration

Alpine Edge
Musl 1.2.3
installer version 6.0.102 with runtime 6.0.2  (as problematic files havn't changed between 6.0.2 and 6.0.3, this issue should exist with current version)

### Other information

_No response_

### PR DESCRIPTION
**Pull request by ayakael**
Fixes https://github.com/dotnet/runtime/issues/67763

### CODE REVIEW

File: src/coreclr/pal/src/init/pal.cpp [hunk 1]
```diff
@@ -120,7 +120,7 @@ BOOL g_useDefaultBaseAddr = FALSE;
 
 /* critical section to protect access to init_count. This is allocated on the
    very first PAL_Initialize call, and is freed afterward. */
-static PCRITICAL_SECTION init_critsec = NULL;
+static PCRITICAL_SECTION init_critsec = nullptr;
```
Comment (am11) [inline src/coreclr/pal/src/init/pal.cpp:123]: Reverting this line should fix the CG2 error.

Comment (ayakael) [reply src/coreclr/pal/src/init/pal.cpp]: pushed

File: src/coreclr/pal/src/include/pal/palinternal.h [hunk 2]
```diff
@@ -695,30 +695,30 @@ T* InterlockedCompareExchangePointerT(
 template <typename T>
 inline T* InterlockedExchangePointerT(
     T* volatile * target,
-    int           value) // When NULL is provided as argument.
+    std::nullptr_t           value) // When NULL is provided as argument.
 {
     //STATIC_ASSERT(value == 0);
-    return InterlockedExchangePointerT(target, reinterpret_cast<T*>(value));
+    return InterlockedExchangePointerT(target, value);
```
Comment (am11) [inline src/coreclr/pal/src/include/pal/palinternal.h:701]:
```suggestion
    return InterlockedExchangePointerT(target, (T*)(void*)value);
```

File: src/coreclr/pal/src/include/pal/palinternal.h [hunk 3]
```diff
@@ -695,30 +695,30 @@ T* InterlockedCompareExchangePointerT(
...
 template <typename T>
 inline T* InterlockedCompareExchangePointerT(
     T* volatile * destination,
-    int           exchange,  // When NULL is provided as argument.
+    std::nullptr_t           exchange,  // When NULL is provided as argument.
     T*            comparand)
 {
     //STATIC_ASSERT(exchange == 0);
-    return InterlockedCompareExchangePointerT(destination, reinterpret_cast<T*>(exchange), comparand);
+    return InterlockedCompareExchangePointerT(destination, exchange, comparand);
```
Comment (am11) [inline src/coreclr/pal/src/include/pal/palinternal.h:711]:
```suggestion
    return InterlockedCompareExchangePointerT(destination, (T*)(void*)exchange, comparand);
```

File: src/coreclr/pal/src/include/pal/palinternal.h [hunk 4]
```diff
@@ -695,30 +695,30 @@ T* InterlockedCompareExchangePointerT(
...
 template <typename T>
 inline T* InterlockedCompareExchangePointerT(
     T* volatile * destination,
     T*            exchange,
-    int           comparand) // When NULL is provided as argument.
+    std::nullptr_t           comparand) // When NULL is provided as argument.
 {
     //STATIC_ASSERT(comparand == 0);
-    return InterlockedCompareExchangePointerT(destination, exchange, reinterpret_cast<T*>(comparand));
+    return InterlockedCompareExchangePointerT(destination, exchange, comparand);
```
Comment (am11) [inline src/coreclr/pal/src/include/pal/palinternal.h:721]:
```suggestion
    return InterlockedCompareExchangePointerT(destination, exchange, (T*)(void*)comparand);
```

**Review by am11 — COMMENTED** [review]
Tested it locally on macOS, this fixes the innerloop build.

**Comment by ayakael** [comment]
@am11 Perfect, that did it. I'm seeing success on Alpine pipelines. Many thanks!

**Comment by am11** [comment]
@ayakael, thanks for adding musl-libc v1.2.3 support early in its lifecycle! 👍

One unique thing about CoreCLR PAL is that it is essentially implementing Windows APIs on Unix, which rest of the code under `src/coreclr` (incl. JIT) uses on Unix. For instance, in this case, it is the implementation of this win32 API: https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedexchangepointer.

The way palinternal is written is it undefines then redefines the low level functions (like those belonging to `printf()` family) which normally come from standard C/C++ toolchain. This is to harmonize the behavior of known APIs with their Windows counterparts, so rest of the code under src/coreclr don't have to care about platform or API behavior differences. Therefore, sometimes you may notice slight mismatches between PAL implementation and the standard C/C++ ones.

File: src/coreclr/pal/src/init/pal.cpp [hunk 5]
```diff
@@ -1139,7 +1139,7 @@ static BOOL INIT_IncreaseDescriptorLimit(void)
 
 Parameters :
     int argc : number of arguments in argv
-    char **argv : argument list in an array of NULL-terminated strings
+    char **argv : argument list in an array of nullptr-terminated strings
```
Comment (janvorli) [inline src/coreclr/pal/src/init/pal.cpp:1142]: A nit - this should be reverted.

File: src/coreclr/pal/src/init/pal.cpp [hunk 6]
```diff
@@ -1221,32 +1221,32 @@ static LPWSTR INIT_FormatCommandLine (int argc, const char * const *argv)
         }
         *command_ptr++=' ';
     }
-    /* replace the last space with a NULL terminator */
+    /* replace the last space with a nullptr terminator */
```
Comment (janvorli) [inline src/coreclr/pal/src/init/pal.cpp:1224]: A nit - this should be reverted

**Review by janvorli — APPROVED** [review]
LGTM modulo the two nits.

**Comment by janvorli** [comment]
@ayakael thank you for the fix!

File: src/coreclr/pal/src/init/pal.cpp (hunk 5 above)
Comment (ayakael) [reply src/coreclr/pal/src/init/pal.cpp]: done :)

File: src/coreclr/pal/src/init/pal.cpp (hunk 6 above)
Comment (ayakael) [reply src/coreclr/pal/src/init/pal.cpp]: done :)

**Comment by ayakael** [comment]
> @ayakael thank you for the fix!

you're welcome ^^ 



> @ayakael, thanks for adding musl-libc v1.2.3 support early in its lifecycle! +1
> 
> One unique thing about CoreCLR PAL is that it is essentially implementing Windows APIs on Unix, which rest of the code under `src/coreclr` (incl. JIT) uses on Unix. For instance, in this case, it is the implementation of this win32 API: https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedexchangepointer.
> 
> The way palinternal is written is it undefines then redefines the low level functions (like those belonging to `printf()` family) which normally come from standard C/C++ toolchain. This is to harmonize the behavior of known APIs with their Windows counterparts, so rest of the code under src/coreclr don't have to care about platform or API behavior differences. Therefore, sometimes you may notice slight mismatches between PAL implementation and the standard C/C++ ones.

many thanks for sharing what PAL is for ^^ Being a packager first, programmer after, means that I mostly throiw spaguetti at a wall to see if it sticks. It's nice to be able to have a background understanding on why things are as they are.

**Comment by nekopsykose** [comment]
this pull request (https://gitlab.alpinelinux.org/alpine/aports/-/commit/2f56551abc53b818bab9f22c4f7972b5ad5aad66 , just missing the two nits) applied onto our alpine build fixed it, and it works again :) (just notifying, since it was alpine that ran into this first)

**Comment by ayakael** [comment]
> this pull request (https://gitlab.alpinelinux.org/alpine/aports/-/commit/2f56551abc53b818bab9f22c4f7972b5ad5aad66 , just missing the two nits) applied onto our alpine build fixed it, and it works again :) (just notifying, since it was alpine that ran into this first)

Many thanks, I had forgot to update.

**Pull request merged**